### ISSUE TITLE
Adding more Optimizers and Losses

### ISSUE THREAD

**Issue by Shashi456** [tensorflow/swift-apis#127]
Just a small-ish roadmap to different Optimizers and losses we can look at to add :

Optimizers:- 
- [ ] Adam
- [ ]
- [ ]
- [ ]

Losses :-

**Comment by rxwei** [comment]
Thanks for listing these!

**Comment by Shashi456** [comment]
@rxwei just a small roadmap, so that we have in mind what we need to add.

**Comment by liberium** [comment]
Maybe it would be better to organise this as a project? It's not a pull-request by nature.

**Comment by Shashi456** [comment]
@liberium. This is not priority right now. I wrote this task list so that we know what we have and what we don't. We are currently implementing all the layers. This is just a simple checklist so that we know where we are, and if someone wants to implement they can help us. Issues are a good way to get open source contributions.

**Comment by Shashi456** [comment]
@dan-zheng @rxwei so this Google group thread [here](https://groups.google.com/a/tensorflow.org/forum/m/#!topic/swift/nIVNOJ1RDYA) . Talks about regularization. Although we are a little far from that stage right now. I think one functionality we should look at adding is retrieving model weights. Is there an easy way to add this ? Could we start looking at this I would definitely like to help in adding this functionality.

**Comment by eaplatanios** [comment]
@Shashi456 I believe that regularization schemes are heavily intertwined with the training abstractions, which do not yet exist in this repo. It would be nice to start a discussion around them at some point though.

**Comment by jon-tow** [comment]
@Shashi456 Categorical Cross-Entropy seems to already be implemented through Softmax Cross-Entropy with Logits. Maybe we can cross it off the list?

**Comment by Shashi456** [comment]
@rxwei So I'm almost done with adding all the losses. Thanks to @Descartess and @jon-tow for helping out in that matter. One of the only listed losses remaining is #31 which is the triplet loss, If you merge it I can add a test for it. 

cc : @dan-zheng @eaplatanios

**Comment by Shashi456** [comment]
@eaplatanios is there any optimizer which hasn't been refactored yet?

**Comment by eaplatanios** [comment]
All that can be refactored currently, have been. We are missing some functionality that would allow us to simplify AdaMax that is not straightforward to add so that’s the only one using key paths right now.

**Comment by iyaja** [comment]
Have you considered adding the [LAMB optimizer](https://arxiv.org/abs/1904.00962) to the list? It seems to be one of the fastest so far (especially with large batches).

**Comment by Shashi456** [comment]
@jph00, thanks :). Will start contributing some optimizers there as well.

**Comment by vbbphc** [comment]
#564 Add Rectified Adam Optimizer

**Comment by Kshitij09** [comment]
[LabelSmoothingCrossEntropy](https://arxiv.org/abs/1512.00567) Loss, originally proposed by Szegedy et al. and further expounded by Geoffrey Hinton et al. in [When does Label Smoothing Help?](https://arxiv.org/abs/1906.02629) have been proven effective strategy for building robust models. While it's relatively simple to integrate, I'd like to contribute this one to get acquainted with _swift for tensorflow!_

**Comment by Kshitij09** [comment]
I've few things to discuss regarding the current implementation of `softmaxCrossEntropy`:

1. I found inconsistencies among docs of tensorflow's [sparse_softmax_cross_entropy](https://www.tensorflow.org/api_docs/python/tf/nn/sparse_softmax_cross_entropy_with_logits) and swift-api's [softmaxCrossEntropy(logits:labels:reduction)](https://www.tensorflow.org/swift/api_docs/Functions#softmaxcrossentropylogits:labels:reduction:). The reason for this comparison is the effective call being made by this [function](https://github.com/tensorflow/swift-apis/blob/90719e6aa2005a7f30ec6593d4e88826af75029a/Sources/TensorFlow/Loss.swift#L237).

As per tensorflow's doc, `logits` are `unscaled log probabilites` whereas swift-apis say they're `One-hot encoded outputs from a neural network`. Also, there's one warning in tensorflow docs
**WARNING:** This op expects unscaled logits, since it performs a softmax on logits internally for efficiency. Do not call this op with the output of softmax, as it will produce incorrect results.

1. As it is mentioned in the docs of [softmaxCrossEntropy](https://www.tensorflow.org/swift/api_docs/Functions#softmaxcrossentropylogits:probabilities:reduction:), _"Each row must be valid probability distribution"_, however, there's no such precondition in the current implementation.

1. Adhering to `tf.compat.v1` implementation of `CategoricalCrossentropy` loss
```python
tf.losses.softmax_cross_entropy(
    onehot_labels,
    logits,
    weights=1.0,
    label_smoothing=0,
    scope=None,
    loss_collection=tf.GraphKeys.LOSSES,
    reduction=Reduction.SUM_BY_NONZERO_WEIGHTS
)
```

We could extend current implementation of `softmaxCrossEntropy` with `label_smoothing` and `weights`, allowing us to create both `Weighted` as well as `LabelSmoothingCrossEntropy`. I'd like to make above mentioned changes and seeking for your thoughts on the same. I've done some [experiments](https://github.com/Kshitij09/swift-experiments/blob/master/explore_softmax_cross_entropy.ipynb) with raw and one-hot encoded inputs and I found that loss varies drastically.

cc: @dan-zheng @eaplatanios
